amplipi/statcollector.py

```diff
@@ -0,0 +1,224 @@
+#!/usr/bin/env python3
+"""Stat Collector for AmpliPi; collects data, saves it to a json file, then phones home with user consent every X days/weeks"""
+import os
+import json
+import re
+import subprocess
+from datetime import datetime, timezone, timedelta
+from shutil import disk_usage
+import requests
+# pylint: disable=no-name-in-module
+from pydantic import BaseModel
+from psutil import virtual_memory
+
+path = "/var/lib/UsageSurvey.json"
+
+def find_matches(list1: list, list2: list):
+  """Takes in two lists, returns a list that only contains data shared by both lists"""
+  set1 = {json.dumps(item, sort_keys=True) for item in list1}
+  set2 = {json.dumps(item, sort_keys=True) for item in list2}
+
+  matches = [json.loads(item) for item in set1 & set2]
+  return matches
+
+
+def average(running_average, weight, new_entry):
+  """
+    Calculates an average using three variables:
+    running_average, the average so far
+    weight, the count of numbers that went into that average
+    new_entry, the newest addition to the dataset
+    Returns the new average
+  """
+  # Round to nearest whole number, then int() to shave of trailing .0
+  return int(round(((running_average * weight) + new_entry) / (weight + 1), 0))
+
+
+class StreamUsageSchema(BaseModel):
+  """Schema for individual stream type usage statistics"""
+  # For the purpose of the below comments, a "survey cycle" is defined as how often the device phones home
+  active_streams: list = []
+
+  # The highest number of concurrent streams running within the survey cycle
+  average_concurrent_streams: int = 0
+  peak_concurrent_streams: int = 0
+
+  # The length in seconds that a stream of this type has been connected within the survey cycle
+  # Counts as long as at least one playing stream of the given type persists between polling calls
+  current_runtime: int = 0
+  average_runtime: int = 0
+  runtime_weight: int = 0
+  peak_runtime: int = 0
+
+  def consume(self, source_info: list, poll_count: int, timediff: timedelta):
+    """Consumes a list of source["info"] json blobs and updates stream statistics tracker with information found within"""
+    stream_names = [item["name"] for item in source_info]
+    stream_count = len(stream_names)
+
+    if stream_count == 0 and len(self.active_streams) != 0:  # If a stream has just been shut off, record average based on current runtime
+      new_average = average(self.average_runtime, self.runtime_weight, self.current_runtime)
+      self.average_runtime = new_average
+      self.runtime_weight += 1
+
+    matches = find_matches(self.active_streams, stream_names)
+    if len(matches) != 0:  # If there is at least one playing match, calculate runtime
+      self.current_runtime = self.current_runtime + timediff.seconds
+
+      self.peak_runtime = max(self.peak_runtime, self.current_runtime) if self.current_runtime < 86400 else 86400
```

Comment:
There's no real reason to limit this to 24 hours, I just had the feeling that more than 24 hours would lead to outliers and this is a method of negating that. It might be worthwhile to remove the limit just so we can see what sort of lifespan that long-lived stream errors depend on